File: csharp/1478.allocate-mailboxes.cs
/*
 * @lc app=leetcode id=1478 lang=csharp
 *
 * [1478] Allocate Mailboxes
 */

using System;

// @lc code=start
public class Solution {

    private static int MinDistance(int[,] cost, int n, int k, int i, int?[,] memo) {
        if (i == n) {
            return k == 0 ? 0 : 2000000;
        }
        if (k < 0) {
            return 2000000;
        }
        if (memo[k, i].HasValue) {
            return memo[k, i].Value;
        }
        int ret = 2000000;
        for (int j = i; j < n; ++j) {
            ret = Math.Min(ret, cost[i, j] + MinDistance(cost, n, k - 1, j + 1, memo));
        }
        memo[k, i] = ret;
        return ret;
    }

    public int MinDistance(int[] houses, int k) {
        int n = houses.Length;
        int[,] cost = new int[n, n];
        Array.Sort(houses);
        for (int i = 0; i < n; ++i) {
            for (int j = i; j < n; ++j) {
                int p = houses[(i + j) / 2];
                for (int t = i; t <= j; ++t) {
                    cost[i, j] += Math.Abs(p - houses[t]);
                }
            }
        }
        return MinDistance(cost, n, k, 0, new int?[k + 1, n]);
    }
}
// @lc code=end

File: java/1478.allocate-mailboxes.java
/*
 * @lc app=leetcode id=1478 lang=java
 *
 * [1478] Allocate Mailboxes
 */

import java.util.Arrays;

// @lc code=start

class Solution {

    private static final int IMPOSSIBLE = 1234567;

    public int minDistance(int[] houses, int k) {
        Arrays.sort(houses);
        int n = houses.length;
        int[][] cost = new int[n][n];
        for (int i = 0; i < n; ++i) {
            for (int j = i + 1; j < n; ++j) {
                int m = houses[(i + j) / 2];
                for (int l = i; l <= j; ++l) {
                    cost[i][j] += Math.abs(m - houses[l]);
                }
            }
        }
        return minDistance(cost, k, n, 0, new Integer[k + 1][n]);
    }

    private static int minDistance(int[][] cost, int k, int n, int i, Integer[][] memo) {
        if (k < 0) {
            return IMPOSSIBLE;
        }
        if (n == i) {
            return k == 0 ? 0 : IMPOSSIBLE;
        }
        Integer m = memo[k][i];
        if (m != null) {
            return m;
        }
        int ret = IMPOSSIBLE;
        for (int ni = i; ni < n; ++ni) {
            ret = Math.min(ret, cost[i][ni] + minDistance(cost, k - 1, n, ni + 1, memo));
        }
        memo[k][i] = ret;
        return ret;
    }
}
// @lc code=end

